### PR TITLE
Add dependency post

### PR DESCRIPTION
**Pull request by jimhester**
This is still somewhat rough, I need to find a photo, but it has been on the back burner for 6 months now, so wanted to get it out there.

### CODE REVIEW

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+I feel this is a major mistake for most projects, and will instead increase
+maintenance, lower functionality and introduce additional bugs.
+
+## dependencies are not equal
+
+One issue with striving for minimal dependencies is it treats all dependencies
+as equivalent. However in reality this is far from the case. There are many
+axis upon which dependencies can differ, but some of the most important include
+
+1. The number of additional dependencies. For example the
```
Comment (jennybc) [inline content/articles/2019-05-itdepends.Rmarkdown]: When you say "additional" dependencies here, at first I thought you were getting at "marginal addition". You're not -- I see that now.

But the marginal angle is one you might consider too. "If I depend on package X, what actually changes for me and my users?" Various scenarios:

  * Package X is already in my dependencies, implicitly, so making it explicit doesn't really change anything.
    - Depending on a Recommended package that ships with R.
    - Depending on a hard dependency (stringi) of one of your hard dependencies (stringr). Yes, technically, this could change but how likely is that and on what time scale.
  * Package X is, empirically speaking, already installed on 90% of my users' machines, due to patterns of actual package co-usage. So making it an explicit dependency causes very few new installations.

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+## prefer a balanced weight
+
+Instead of striving for absolutely minimal dependencies I would instead suggest
+a more holistic, balanced and quantitative approach.
+
+A holistic approach looks at the project as a whole and asks who is the primary
+audience? If the audience is other package authors than a leaner package with fewer
+dependencies may be more appropriate. If instead it is primarily data scientists or
+statisticians they will likely already have many popular dependencies installed and
+a more feature-full package is preferred.
```
Comment (jennybc) [inline content/articles/2019-05-itdepends.Rmarkdown]: Also: consistency of features. I think of tidy eval and the name repair stuff as examples of us trying to make it increasingly impossible for different packages to offer NSE and name repair, but to implement them in slightly different ways by accident.

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+Then, use the `dep_locate()` function to find all the instances when you use
+functions from a given dependency in your code base. If `dep_locate()` is run
+in RStudio the function will add source markers for all the locations, which
+can then be clicked individually to do the replacement.
+
+```{r}
+itdepends::dep_locate("purrr", path = "~/p/tidyversedashboard")
+```
+
+## Be balanced and use data
```
Comment (jennybc) [inline content/articles/2019-05-itdepends.Rmarkdown]: The idea of using data might be interesting to connect back to profiling and its importance for tuning performance.

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+In almost any area of life striving towards absolutes rarely provides the best
+outcome. Package dependencies are no exception. Keep the expected users of your
+package in mind and use the tools provided in itdepends to obtain data which
```
Comment (batpigandme) [inline content/articles/2019-05-itdepends.Rmarkdown]:
```suggestion
package in mind, and use the tools provided in itdepends to obtain data which
```

**Comment by hadley** [comment]
Photo ideas:

* hug: https://unsplash.com/photos/vzFTmxTl0DQ
* dog pack: https://unsplash.com/photos/U6nlG0Y5sfs
* spiderweb: https://unsplash.com/photos/jpgRztEuaV4
* sheep: https://unsplash.com/photos/24bzOuENxHc 🤣 
* cables: https://unsplash.com/photos/l090uFWoPaI

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+  fig.asp = 0.618, # 1 / phi
+  out.width = "700px"
+)
+```
+
+# It depends - A dialog about dependencies
+
+> This post paraphrases my [talk with the same name](https://resources.rstudio.com/rstudio-conf-2019/it-depends-a-dialog-about-dependencies) at RStudio::conf 2019.
+
+Software dependencies can often be a double edged sword.
```
[the remark this answers is not in the document]
Comment (hadley) [reply content/articles/2019-05-itdepends.Rmarkdown]:
```suggestion
Software dependencies can be a double edged sword.
```

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+marvelous new features and reducing bugs.
+
+On the other hand they can change over time, causing your software to break
+unexpectedly and (potentially) increasing your maintenance burden. They can also
+ take additional disk space and installation time.
+
+
+These downsides have led some to suggest a 'dependency zero' mindset.
+
+I feel this is a major mistake for most projects, and will instead increase
```
Comment (hadley) [inline content/articles/2019-05-itdepends.Rmarkdown]:
```suggestion
We feel that this is a major mistake for most projects, and will instead increase
```

File: content/articles/2019-05-itdepends.Rmarkdown (hunk 1 above)
Comment (hadley) [reply content/articles/2019-05-itdepends.Rmarkdown]: Maybe "downstream" dependencies would be a better term here? 

And @jennybc point deserves discussion somewhere.

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+   For instance the [rgdal](https://cran.r-project.org/package=rgdal) package
+   requires the GDAL and PROJ.4 dependencies, each of which in turn have
+   additional system dependencies. Another example is
+   [rJava](https://cran.r-project.org/package=rJava) which requires Java and
+   also has additional steps required to configure R for the proper Java
+   installation, which has caused [issues for many
+   people](https://stackoverflow.com/questions/tagged/rjava).
+
+I hope the above examples illustrate that not all package dependencies are
+equivalent and we should not treat them as such by using a simple count of
```
Comment (hadley) [inline content/articles/2019-05-itdepends.Rmarkdown]:
```suggestion
equal, and we should not treat them as such by using a simple count of
```

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+dependencies as our metric.
+
+## prefer a balanced weight
+
+Instead of striving for absolutely minimal dependencies I would instead suggest
+a more holistic, balanced and quantitative approach.
```
Comment (hadley) [inline content/articles/2019-05-itdepends.Rmarkdown]:
```suggestion
a more holistic, balanced, and quantitative approach.
```

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+A holistic approach looks at the project as a whole and asks who is the primary
+audience? If the audience is other package authors than a leaner package with fewer
```
Comment (hadley) [inline content/articles/2019-05-itdepends.Rmarkdown]:
```suggestion
audience? If the audience is other package authors then a leaner package with fewer
```

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+underestimate the introduction of new bugs.
+
+However even if you truly _are_ a more skilled programmer than the maintainer of
+a dependency I would still advise caution. The existing package has the
+significant benefits of _time_ and _users_. Users who have undoubtedly uncovered
+numerous bugs using the package in the real world, which are now fixed.
+
+## itdepends, a package to help with dependencies
+
+The [itdepends](https://github.com/jimhester/itdepends) package provides tools
```
Comment (hadley) [inline content/articles/2019-05-itdepends.Rmarkdown]: Move to r-lib? Add pkgdown site?

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+to assess usage, measure weights, visualize proportions and if needed, assist
+removal of dependencies.
+
+## Assess usage with `itdepends::dep_usage()`
+
+`dep_usage_proj()` and `dep_usage_pkg()` return a tibble of
+every function call in your project or package. This information can then be
```
Comment (hadley) [inline content/articles/2019-05-itdepends.Rmarkdown]: I feel like this needs a footnote — I assume it isn't actually every call, but every call that you can find with some straightforward static analysis.

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+used to do things like
+
+- show which packages you are using the most in a project
+
+```{r}
+library(tidyverse)
+itdepends::dep_usage("~/p/tidyversedashboard") %>%
```
Comment (hadley) [inline content/articles/2019-05-itdepends.Rmarkdown]: Might be a worth a `create_from_github()` to make more reproducible?

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+- time to compile dependencies, and the raw times (`install_self`, `install_user`, `install_dev`, `self_timings`, `user_timings`, `dev_timings`)
+- CRAN information (`downloads`, `first_release`, `last_release`, `total_releases`, `releases_last_52`)
+
+The total number of dependencies, binary sizes and installation times give you
+a good idea the magnitude of the dependency load.
+
+Equally important however are the downloads, release information and open
+issues, which tell you how long the package has existed, how many people are
+using it and how many open issues exist. Packages that have existed longer,
+with more users and fewer issues should be the most reliable.
```
Comment (hadley) [inline content/articles/2019-05-itdepends.Rmarkdown]:
```suggestion
with more users and fewer issues should generally be more reliable.
```

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+
+## Assist removal with `itdepends::dep_locate()`
+
+If you do decide that a given dependency is too heavy or is too unstable to
+keep, what are the best practices for removing it?
+
+First and most importantly, write _tests_ for the existing
+behavior. This is an essential step which will greatly reduce the introduction
+of new bugs. Only once you have written the tests should start performing the
+replacement.
```
Comment (hadley) [inline content/articles/2019-05-itdepends.Rmarkdown]: Carefully review the operation of the function that you're replacing?  I want to say something here like: many tidyverse functions appear superficially similar to base equivalents, but they often handle edge cases (that you've possibly never seen) more consistently. But without tying it so strongly to tidyverse.

File: content/articles/2019-05-itdepends.Rmarkdown (hunk 1 above)
Comment (jimhester) [reply content/articles/2019-05-itdepends.Rmarkdown]: Yes agreed on this being something we should mention

File: content/articles/2019-05-itdepends.Rmarkdown (hunk 10 above)
Comment (jimhester) [reply content/articles/2019-05-itdepends.Rmarkdown]: Probably needs a better readme as well...

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+
+A quantitative approach lets you more accurately measure your dependencies, and
+see which are heavier or lighter along various axes. Measuring each of the
+dependencies gives you more power to reason about which dependencies would be
+most beneficial to remove (if any).
+
+How you weigh these costs and benefits varies by the skills and experience of
+the maintainer and should also vary by the goals of the project.
+
+## beware illusionary superiority
```
[the remark this answers is not in the document]
Comment (jimhester) [reply content/articles/2019-05-itdepends.Rmarkdown]: 🙊

File: content/articles/2019-05-itdepends.Rmarkdown (hunk 14 above)
Comment (jennybc) [reply content/articles/2019-05-itdepends.Rmarkdown]: Maybe something like:

If you re-implement a function, you should probably port over its tests, to make sure you're handling edge cases the same way.

File: content/articles/2019-05-itdepends.Rmarkdown (hunk 11 above)
Comment (jimhester) [reply content/articles/2019-05-itdepends.Rmarkdown]: It is really every call, including things like `{` and `<-` etc., I can't really think of a call that wouldn't be represented, but it is R, so it is probably possible to construct one.

File: content/articles/2019-05-itdepends.Rmarkdown (hunk 12 above)
Comment (jimhester) [reply content/articles/2019-05-itdepends.Rmarkdown]: I switched this to use `dep_usage_pkg()` so it is reproducible across environments and doesn't depend on a specific project.

**Comment by jimhester** [comment]
Thank you for the reviews @batpigandme, @jennybc and @hadley, I think they improved the post a lot!

I made some moderate changes incorporating the suggestions, if someone could do a quick read through to ensure I didn't accidentally mess anything up that would be appreciated. Probably the largest change was the conclusion, so thoughts there would be especially useful!

File: content/articles/2019-05-itdepends.Rmarkdown (hunk 11 above)
Comment (hadley) [reply content/articles/2019-05-itdepends.Rmarkdown]: I meant things like `lapply(x, foo)` — it's not going to pick up `foo`.

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+  [drat](https://github.com/eddelbuettel/drat) which require users to configure
+  additional repositories before installation.
+
+2. The number of downstream dependencies. For example, the
+   [rlang](https://github.com/r-lib/rlang) package has no downstream
+   dependencies apart from R; while the
+   [CNVscope](https://cran.r-project.org/package=CNVScope) package has 125 downstream
+   dependencies, on both CRAN and Bioconductor.
+
+3. Already fulfilled dependencies. If your package depends on `dplyr` than
```
Comment (jennybc) [inline content/articles/2019-05-itdepends.Rmarkdown]:
```suggestion
3. Already fulfilled dependencies. If your package depends on `dplyr` then
```

File: content/articles/2019-05-itdepends.Rmarkdown [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+
+A balanced approach understands that adding (or removing) dependencies
+come with trade-offs. Adding a dependency gives you additional features,
+bug fixes, and real-world testing, at the costs of increased installation time,
+disk space and maintenance if the dependency has breaking changes. In some
+cases it makes sense to _increase_ dependencies for a package, even if an
+implementation already exists. For instance base R has a number of different
+implementations of non-standard evaluation with varying semantics across its
+functions. The same used to be true of tidyverse packages as well, but now they
+ages all depend on the implementations in the tidyselect and rlang packages,
```
Comment (jennybc) [inline content/articles/2019-05-itdepends.Rmarkdown]:
```suggestion
all depend on the implementations in the tidyselect and rlang packages,
```

File: content/articles/2019-05-itdepends.Rmarkdown (hunk 11 above)
Comment (jimhester) [reply content/articles/2019-05-itdepends.Rmarkdown]: yeah true, I will add a footnote mentioning this

**Comment by jimhester** [comment]
So the post was actually missing from the preview because I did not commit the `.markdown` file, which I have now done.

Thanks again everyone for all the review comments, I think it is much improved thanks to your feedback!
